### PR TITLE
Fixes Workflow Timer

### LINKED ISSUE #3436: Duplicate record errors in workflows
Hi,

We still have sometimes "database is locked" errors. When this happens, duplicate key errors are blocking Worflows.

Restarting OC does not clear the issue. We have to delete the duplicate entry in WorkflowTypeStartActivitiesIndex.

```
2019-04-05 09:51:47.6767|Default|0HLLPJE52JTCO:00000001||Microsoft.AspNetCore.Server.Kestrel|ERROR|Connection id "0HLLPJE52JTCO", Request id "0HLLPJE52JTCO:00000001": An unhandled exception was thrown by the application. System.ArgumentException: An item with the same key has already been added. Key: 4jpqrrzbq2cnp08g132j6sxecb
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](List`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at OrchardCore.Workflows.Http.Startup.Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider) in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Workflows\Http\Startup.cs:line 76
   at OrchardCore.Modules.ModularTenantRouterMiddleware.ConfigureTenantPipeline(IApplicationBuilder appBuilder, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 125
   at OrchardCore.Modules.ModularTenantRouterMiddleware.<>c__DisplayClass6_0.<BuildTenantPipeline>b__0(IApplicationBuilder builder) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 96
   at Microsoft.AspNetCore.Mvc.Internal.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.BuildTenantPipeline(IServiceProvider rootServiceProvider, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 105
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 71
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 100
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)    at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](List`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at OrchardCore.Workflows.Http.Startup.Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider) in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Workflows\Http\Startup.cs:line 76
   at OrchardCore.Modules.ModularTenantRouterMiddleware.ConfigureTenantPipeline(IApplicationBuilder appBuilder, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 125
   at OrchardCore.Modules.ModularTenantRouterMiddleware.<>c__DisplayClass6_0.<BuildTenantPipeline>b__0(IApplicationBuilder builder) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 96
   at Microsoft.AspNetCore.Mvc.Internal.MiddlewareFilterBuilderStartupFilter.<>c__DisplayClass0_0.<Configure>g__MiddlewareFilterBuilder|0(IApplicationBuilder builder)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.BuildTenantPipeline(IServiceProvider rootServiceProvider, IServiceProvider scopeServiceProvider) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 105
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 71
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 100
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)

```

```
2019-04-05 09:04:51.2594|Default|0HLLP12PLPULC:0000000D||Microsoft.AspNetCore.Server.Kestrel|ERROR|Connection id "0HLLP12PLPULC", Request id "0HLLP12PLPULC:0000000D": An unhandled exception was thrown by the application. Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 6: 'database table is locked'.
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()
   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1226
   at YesSql.Session.UpdateEntityAsync(Object entity)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchard2\src\OrchardCore\OrchardCore.Abstractions\Shell\Builders\ShellContext.cs:line 279
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 118
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)    at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()
   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1226
   at YesSql.Session.UpdateEntityAsync(Object entity)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.Dispose()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()
   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchard2\src\OrchardCore\OrchardCore.Abstractions\Shell\Builders\ShellContext.cs:line 279
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 118
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)

```
image

image

### LINKED ISSUE #5223: Workflows : Single Instance option is not observed when workflow is "Halted"
When creating a workflow type that uses the "Single Instance" property option, the workflow executes multiple instances even though it is not "Finished". This occurs when a workflow is "Halted". For example, adding a Http Request Event to a workflow results in the workflow being Halted until an event comes in and triggers that web hook. If that event does not arrive before the next timer interval, a new workflow is still started.

Example :

[TIMER EVENT(Every minute)] ---> [TASK] -----> [Http Request Event] ----> [TASK]

### LINKED ISSUE #5231: Workflows : Disabled Time Based workflow still executes scheduled workflow
After disabling a workflow where a Timer is the starting Task, the workflow task still executes. This seems to happen since a Workflow Type queues a new Workflow event and the Timer task queues a Halted Workflow item prior to it executing. Then when it comes time to execute the timed event, even though the workflow is disabled, it still executes.

I would argue it should remove this queued item or at least not execute until it is Enabled again.

### LINKED ISSUE #5810: Workflow Timer Event creates X amount of "HALTED" workflows
I have created a Timer Event in a workflow that is only supposed to run once a day at 00:00 using 
CRON: `0 0 * * *`
https://crontab.guru/every-night

However, at every minute when the background task runs, it creates a new HALTED workflow for this workflow type. After a couple of hours i had a lot..

So in order to test it out, I tried creating another Timer Event with
CRON: `*/5 * * * *` to make it easier to see what would happen.
https://crontab.guru/every-5-minutes

Result:
This will create a HALTED workflow every minute. Then, on the fifth event firing, it will create a new workflow as before but then execute all five at the same time.  So either i'm doing something completely wrong or the cron/event task is seriously bugged?

### PR DESCRIPTION
**Pull request by jtkech**
Fixes #5810 
Fixes #5231 
Fixes #5223 

I think that it fixes #3436 

As i commented in the source code: The 'TimerEvent' is a one minute signal and in fact this is the timer activity that checks if it has expired. So, because a timer event can't occur again before it has previously expired, if there is already a workflow instance that is halted on a timer activity, we should not start a new instance, we just have to resume it.

Otherwise, if you have a timer event as a starting activity in a workflow type that is not marked as a singleton. E.g if the timer is set to expire every 5 minutes, you will create 5 instances that will get called, one after the other in the same scope.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs [hunk 1]
```diff
@@ -133,8 +134,12 @@ public async Task TriggerEventAsync(string name, IDictionary<string, object> inp
             // Start new workflows.
             foreach (var workflowType in workflowTypesToStart)
             {
+                // The 'TimerEvent' is a one minute signal and in fact this is the timer activity that checks if it has expired.
+                // So, because a timer event can't occur again before it has previously expired, if there is already a workflow
+                // instance that is halted on a timer activity, we should not start a new instance, we just have to resume it.
+
                 // If this is a singleton workflow and there's already an instance, then skip.
-                if (workflowType.IsSingleton && haltedWorkflows.Any(x => x.WorkflowTypeId == workflowType.WorkflowTypeId))
+                if ((workflowType.IsSingleton || name == TimerEvent.EventName) && haltedWorkflows.Any(x => x.WorkflowTypeId == workflowType.WorkflowTypeId))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:142]: Fat hack. Maybe a custom property in tasks that would describe this behavior instead. Like what if another task needed to have the same behavior?

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:142]: Like `IsExclusive` or something similar

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs:142]: Yes but here there are singularities that i tried to describe in my comments, one being that in fact the timer event is triggered every minute, one being that it is closely tied to the TimerEvent type ...

`IsExclusive` is a good idea, hmm i suggest a const property as it doesn't make sense to make it updatable

I will take a look.

**Comment by jtkech** [comment]
@sebastienros 

Done for the `IsExclusive` property.

So yes, in other words a time event at **a given unique time can't happen twice, this is by nature**, that's why i was checking it here. **I didn't find any other event that would need to behave like this implicitly**, only explicitly by using the `Singleton` property that already exists and is editable.

Note: We could have also used an `isExclusive` param for `TriggerEventAsync()` that would be set to true when called from the timer background, or for any other kind of exclusive event if any.

**Comment by sebastienros** [comment]
> We could have also used an isExclusive param for TriggerEventAsync() 

This sound even better than the property

**Comment by jtkech** [comment]
Okay, i will update it this night

**Comment by jtkech** [comment]
This one is safe as i only add an `isExclusive` param to the `TriggerEventAsync()`

**Pull request merged**